### ISSUE TITLE
LTS brings in previously known broken versions

### ISSUE THREAD

**Issue by exarkun** [emilypi/Base32#11]
Here's an example of a ghci session where encodeBase32Unpadded returns several different outputs for the same input:

```
*Main CHK Crypto Lib Types Upload> import Data.ByteString.Base32.Hex
*Main CHK Crypto Lib Types Upload Data.ByteString.Base32.Hex> let a = "\n&\250\240\218B\n\t\225[\137\222a\253d\206"
*Main CHK Crypto Lib Types Upload Data.ByteString.Base32.Hex> encodeBase32Unpadded a
"POJFLS6Q8850JOARH7F63VB4y\ACK"
*Main CHK Crypto Lib Types Upload Data.ByteString.Base32.Hex> encodeBase32Unpadded a
"*** Exception: Cannot decode byte '\x91': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
*Main CHK Crypto Lib Types Upload Data.ByteString.Base32.Hex> encodeBase32Unpadded a
"POJFLS6Q8850JOARH7F63VB4\SOH\NUL"
*Main CHK Crypto Lib Types Upload Data.ByteString.Base32.Hex> encodeBase32Unpadded a
"*** Exception: Cannot decode byte '\x92': Data.Text.Internal.Encoding.decodeUtf8: Invalid UTF-8 stream
```

The two non-exception results are different and neither is completely base32.  Of the exception cases, the exception mentions a different byte each time.

Note the exception cases seem much, much more common than the non-exception cases.  Try it 20 or 30 times if it only seems to be hitting the exception.  Even the exception result seems like a bug to me though - encoding should always succeed, probably?  Or it should at least always fail with the same exception.

This is on ghc 8.8.4, NixOS git rev ff6fda61600cc60404bab5cb6b18b8636785b7bc, x86_64.

**Comment by emilypi** [comment]
Thanks for raising this. I'll need to find a way to repro and investigate. On my side, everything looks fine: 

```haskell
П> import qualified Data.ByteString.Base32.Hex as H
П> let a = "\n&\250\240\218B\n\t\225[\137\222a\253d\206"
П> H.encodeBase32Unpadded a
"18JFLS6Q8850JOARH7F63VB4PO"
```

The odd thing about your encoding is that it looks like it's not encoding the tail-end correctly -at all-. I remember this being an issue for certain inputs in the last major version for unpadded encodings (subsequently, the test suite is now _massive_ to catch these cases - I'm suprised a new one popped up!). Could you confirm the version of base32 that Nix is pulling in? I wonder if it's pinned to an older version, since nixpkgs relies on Stackage for its curation.

**Comment by exarkun** [comment]
I'm using stack with lts-16.26 resolver - so it looks like base32 0.1.1.2 (which I just looked up on stackage.org; if there's a more direct way to find the version stack is using I'm not sure what it is).

So it sounds like you think this is fixed in 0.2?

**Comment by exarkun** [comment]
I pulled base32 0.2.0.0 into the lts-16.26 stack and the problem seems to have gone away.  I now consistently get the behavior in https://github.com/emilypi/Base32/issues/11#issuecomment-751418695

**Comment by emilypi** [comment]
> if there's a more direct way to find the version stack is using I'm not sure what it is).

There is not unfortunately. This is one of the downsides with curation: when a major version fix comes in, the old LTS's tie you to the buggy version indefinitely until you upgrade. I'm going to open an issue with them and see if they can't backport a fix to the older LTS's updating the version bounds on base32. 

> I pulled base32 0.2.0.0 into the lts-16.26 stack and the problem seems to have gone away. I now consistently get the behavior in #11 (comment)

phew. 😅 - alright, I'm glad it's working. Mild heart attack there. I'll leave the issue open until we resolve the LTS problem.

**Comment by emilypi** [comment]
@exarkun okay, 2.0 will be included in LTS 16.28 and going forward thanks to the Stackage folks: https://www.stackage.org/lts-16.28/package/base32-0.2.0.0. I'm going to close this as "fixed" - thanks for raising it.
